packages/salesforcedx-vscode-core/src/commands/commands.ts

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (c) 2017, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import {
+  CliCommandExecutor,
+  Command
+} from '@salesforce/salesforcedx-utils-vscode/out/src/cli';
+import * as path from 'path';
+import * as vscode from 'vscode';
+import { channelService } from '../channels';
+import { notificationService } from '../notifications';
+import { isSfdxProjectOpened } from '../predicates';
+import { CancellableStatusBar, taskViewService } from '../statuses';
+
+// Precondition checking
+////////////////////////
+export interface PreconditionChecker {
+  check(): boolean;
+}
+
+export class SfdxWorkspaceChecker implements PreconditionChecker {
+  public check(): boolean {
+    const result = isSfdxProjectOpened.apply(vscode.workspace);
+    if (!result.result) {
+      notificationService.showErrorMessage(result.message);
+      return false;
+    }
+    return true;
+  }
+}
+
+// Input gathering
+//////////////////
+export interface ContinueResponse<T> {
+  type: 'CONTINUE';
+  data: T;
+}
+
+export interface CancelResponse {
+  type: 'CANCEL';
+}
+
+export interface ParametersGatherer<T> {
```

Comment:
When we gather parameters, the user can either provide it or cancel out of it.
Instead of overloading `null` to signify a cancellation, I distinguish it using a Continue | CancelResponse.
This is similar to the idea of an Either type in Scala or the tuple return type in Go.